src/scanner.c

```diff
@@ -0,0 +1,310 @@
+#include <assert.h>
+#include <string.h>
+#include <tree_sitter/parser.h>
+#include <wctype.h>
+
+enum TokenType { COMMENT };
+
+typedef struct {
+  size_t quoted_string_id_length;
+  size_t quoted_string_id_capacity;
+  char *quoted_string_id;
+} Scanner;
+
+static inline void quoted_string_id_clear(Scanner *scanner) {
+  scanner->quoted_string_id_length = 0;
+}
+
+static inline void quoted_string_id_resize(Scanner *scanner,
+                                           size_t min_capacity) {
+  size_t capacity = scanner->quoted_string_id_capacity;
+
+  if (capacity < 16) capacity = 16;
+  while (capacity < min_capacity) capacity <<= 1;
+
+  scanner->quoted_string_id_capacity = capacity;
+  scanner->quoted_string_id =
+      realloc(scanner->quoted_string_id, capacity * sizeof(char));
+}
+
+static inline void quoted_string_id_push(Scanner *scanner, char c) {
+  quoted_string_id_resize(scanner, scanner->quoted_string_id_length + 1);
```

Comment:
I don't really like calling `realloc` at every push.

Sure, I get that using glibc's `realloc` proper reallocation only happens when we hit powers of two because of the logic implemented in `quoted_string_id_resize`. But there is nothing enforcing this behaviour in the documentation. So a compliant libc implementation might decide to move the data around in memory even when the requested size is the same as the current size of the data. I would prefer to have an `if` statement in `quoted_string_id_resize` to bypass `realloc` when `capacity == scanner->quoted_string_id_capacity`

Reply:
Good remark. I added a check.